text/3668-async-closure.md

```diff
@@ -0,0 +1,541 @@
+
+### Closures cannot return futures that borrow from their captures
+
+When users wants to *call* a function that takes an async callback argument, they often reach for `|| async {}` (a closure that returns an anonymous future) because closures can capture state from the local environment and are syntactically lightweight.
+
+However, users are quickly met with the limitation that they cannot use any of the closure's captures by reference in the async block.  E.g.:
+
+```rust
+async fn parent_country(city_name: &str) -> String { todo!() }
+
+async fn main() {
+    // Collect the country names of each city in our list.
+    let mut countries = vec![];
+    for_each_city(|city_name| async {
+        countries.push(parent_country(city_name).await);
+    })
+}
+```
+
+```
+error: captured variable cannot escape `FnMut` closure body
+  --> src/main.rs
+   |
+   |       let mut countries = vec![];
+   |           ------------- variable defined here
+   |       for_each_city(|city_name| async {
+   |  _____________________________-_^
+   | |                             |
+   | |                             inferred to be a `FnMut` closure
+   | |         countries.push(parent_country(city_name).await);
+   | |         --------- variable captured here
+   | |     }).await;
+   | |_____^ returns an `async` block that contains a reference to a captured variable, which then escapes the closure body
+   |
+   = note: `FnMut` closures only have access to their captured variables while they are executing...
+   = note: ...therefore, they cannot allow references to captured variables to escape
+```
+
+The future that is returned by the closure cannot reference any of the captures of the closure, which is a limitation that makes `|| async {}` quite unusable today without needing to, for example, clone data and declare the async block `move`.
+
+In order for this to work, the `FnMut` trait would need to be ["lending"](https://smallcultfollowing.com/babysteps/blog/2023/05/09/giving-lending-and-async-closures/#async-closures-are-a-lending-pattern); however, there are [complications](https://hackmd.io/@compiler-errors/async-closures#Lending-closures-are-not-typically-FnOnce) with implementing general lending closures.
+
+# Guide Level Explanation
+
+Just as you can write functions which accept closures, you can write functions which accept async closures:
+
+```rust
+async fn takes_async_closure(f: impl async Fn(u64)) {
+    f(0).await;
+    f(1).await;
+}
+
+takes_async_closure(async |i| {
+    core::future::ready(i).await;
+    println!("done with {i}.");
+});
+```
+
+We recommend using `async Fn` and `async ||` for async closures. This is more flexible than a closure returning a future for the reasons described elsewhere in this RFC.
```

Comment:
The point here is that after `async Fn` and `async ||`, the general recommendation is to grab those out of your toolbox rather than writing `F: Fn() -> Fut, Fut: Future` and `|| async {}`, so it really is a *recommendation* in the sense that the guide should teach users about async closures/fn trait bounds and recommend using them.

We are indeed proposing the existence of the syntax too, I guess, but that's the less important part (since the rest of the RFC kinda assumes that's a given).

Reply:
What this section is saying is that we recommend that people write `async || { .. }` rather than `|| async { .. }`, and that we recommend people write `T: async Fn()` rather than `T: Fn() -> F, F: (Into?)Future<Output = ()>`.